### DISCUSSION TITLE
Adding LocalDB to Aspire

### DISCUSSION THREAD

**Discussion by ChakaTmp** [microsoft/aspire#1656] [Q&A] [answered]
How can I add LocalDb to an Aspire project with my own specific database path?

**Comment by DamianEdwards** [comment, marked as answer]
IIRC LocalDb is auto-activated when a connection attempt is made with the appropriate connection string format. See details at https://learn.microsoft.com/sql/database-engine/configure-windows/sql-server-express-localdb

So to use it in Aspire, you'd simply need to call `WithReference` passing a `ConnectionString` instance with a LocalDB connection string (assume you have it installed), e.g. 

```csharp
builder.AddProject<Projects.ApiService>("apiservice")
    .WithReference(new ConnectionString("LocalDB", "Server=(localdb)\\MSSQLLocalDB;Integrated Security=true"));
```

I verified this works locally.

**Comment by verbedr** [comment]
So this doesn't work any more.  I'm currently doing the following ...

In the AppHost

```
if (!builder.Environment.IsDevelopment())
{
    var sql = builder.AddSqlServer("sql").PublishAsAzureSqlDatabase();
    var sqlDb = sql.AddDatabase("sqldb");
    webFrontend.WithReference(sqlDb);
}
```

In the consuming  webfront app

```
var connectionString = builder.Configuration.GetConnectionString("DefaultConnection");
builder.Services.AddDbContext<DbContext>((provider, builder) =>
{
    builder.UseSqlServer(connectionString);
});
builder.EnrichSqlServerDbContext<DbContext>();
```

I'm just making this comment because I believe the use case is valid.  I don't want my team to be forced to install docker to run a dev sql server when they have it out of the box. An option on the IResourceBuilder<SqlServerServerResource> exists to run AsAzureSqlDatabase. I would think an AsLocalSqlDatabase could be added?

**Reply by davidfowl** [reply]
What's not working exactly?

**Reply by verbedr** [reply]
Sorry, should have added more explanation.

The class Azure.Provisioning.ConnectionString has become abstract.  Using one of the derived ones isn't possible, as their constructors are internal.

https://learn.microsoft.com/en-us/dotnet/api/azure.provisioning.connectionstring?view=azure-dotnet-preview

**Reply by davidfowl** [reply]
This type has nothing to do with local db or using connection strings in aspire locally. Can you share more about what you tried and what doesn't work?

**Reply by verbedr** [reply]
I tried to implement Damian's answer above.  So basically, have something along the lines of this

```csharp
using Azure.Provisioning;
using Microsoft.Extensions.Hosting;

var builder = DistributedApplication.CreateBuilder(args);

var webFrontend = builder.AddProject<Projects.Web>("webfrontend")
    .WithExternalHttpEndpoints();

if (builder.ExecutionContext.IsPublishMode)
{
    var sql = builder.AddSqlServer("sql").PublishAsAzureSqlDatabase();
    var sqlDb = sql.AddDatabase("sqldb");
    webFrontend.WithReference(sqlDb);
}
else
{
    webFrontend.WithReference(new ConnectionString("sqldb", "Server=(localdb)\\MSSQLLocalDB;Integrated Security=true"));
}

builder.Build().Run();
``` 

Or am I overthinking this?

Updated it to incorporate the remark from Ryan

**Reply by davidfowl** [reply]
```C#
var builder = DistributedApplication.CreateBuilder(args);

var webFrontend = builder.AddProject<Projects.WebApplication1>("webfrontend")
    .WithExternalHttpEndpoints();

IResourceBuilder<IResourceWithConnectionString> sqlDb = default;

if (builder.ExecutionContext.IsPublishMode)
{
    sqlDb = builder.AddSqlServer("sql").PublishAsAzureSqlDatabase()
        .AddDatabase("sqldb");
}
else
{
    sqlDb = builder.AddConnectionString("sqldb");
}

webFrontend.WithReference(sqlDb);

builder.Build().Run();
```

Then specify ConnectionStrings:sqlDb in appsettings.json in the apphost (or in user secrets)

**Reply by verbedr** [reply]
Thanks a lot! This works perfectly, getting more and more appreciation for Aspire!

**Comment by ryanheath** [comment]
You probably need to replace

`if (builder.Environment.IsDevelopment())`

with 

`if (!builder.ExecutionContext.IsPublishMode)`

// Ryan

**Reply by verbedr** [reply]
Thanks! Excellent tip.
